platforms/core-runtime/native/src/main/java/org/gradle/internal/nativeintegration/services/NativeServices.java

```diff
@@ -73,64 +75,101 @@
-                            return false;
-                        }
-                    }
-                    try {
-                        FileEvents.init(nativeBaseDir);
-                        LOGGER.info("Initialized file system watching services in: {}", nativeBaseDir);
-                        return true;
-                    } catch (NativeIntegrationUnavailableException ex) {
-                        logFileSystemWatchingUnavailable(ex);
+            public boolean initialize(File nativeBaseDir, ServiceRegistryBuilder builder, boolean useNativeIntegrations) {
+                if (!useNativeIntegrations) {
+                    return false;
+                }
+                OperatingSystem operatingSystem = OperatingSystem.current();
+                if (operatingSystem.isMacOsX()) {
+                    String version = operatingSystem.getVersion();
+                    if (VersionNumber.parse(version).getMajor() < 12) {
+                        LOGGER.info("Disabling file system watching on macOS {}, as it is only supported for macOS 12+", version);
                         return false;
                     }
                 }
+                try {
+                    final FileEvents fileEvents = FileEvents.init(nativeBaseDir);
+                    LOGGER.info("Initialized file system watching services in: {}", nativeBaseDir);
+                    builder.provider(new ServiceRegistrationProvider() {
+                        @Provides
+                        FileEventFunctionsProvider createFileEventFunctionsProvider() {
+                            return new FileEventFunctionsProvider() {
+                                @Override
+                                public <T extends NativeIntegration> T getFunctions(Class<T> type) {
+                                    if (fileEvents != null) {
+                                        return fileEvents.get(type);
+                                    } else {
+                                        throw new NativeIntegrationUnavailableException("File events are not available.");
+                                    }
+                                }
+                            };
+                        }
+                    });
+                    return true;
+                } catch (NativeIntegrationUnavailableException ex) {
+                    logFileSystemWatchingUnavailable(ex);
+                }
                 return false;
             }
+
+            @Override
+            public void doWhenDisabled(ServiceRegistryBuilder builder) {
+                // We still need to provide an implementation of FileEventFunctionsProvider,
+                // even if file watching is disabled, otherwise the service registry will throw an exception for a missing service.
+                builder.provider(new ServiceRegistrationProvider() {
+                    @Provides
+                    FileEventFunctionsProvider createFileEventFunctionsProvider() {
```

Comment:
💭 We need to inject dummy service, even File System watching is disabled, otherwise our DI framework will throw an error that service is missing when creating `BuildLifecycleAwareVirtualFileSystem`